src/net_processing.cpp

```diff
@@ -5194,6 +5197,65 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
     }
 }
 
+void PeerManagerImpl::CheckForLimitedMempoolAndEvictPeers()
+{
+    LOCK(cs_main);
+    // It only makes sense once we're out of IBD, because otherwise our mempool
+    // is not a good source for fee estimation anyway.
+    // TODO: assumeUTXO?
+    if (!CanDirectFetch()) return;
+
+    // Don't bother if we are still seeking for outbound peers: eviction
+    // instead will reduce our chances to get blocks and transactions from
+    // them.
+    // TODO this could be more nuanced.
+    if (m_connman.GetExtraFullOutboundCount() < 0) return;
+
+    // Ideally we want to compare mempool sizes, not fee filters.
+    // Otherwise we easily get confused: e.g. at empty mempools this is less
+    // critical, but that'd be impossible to account for.
+    const auto our_min_feerate = WITH_LOCK(m_mempool.cs, return m_mempool.m_min_relay_feerate.GetFeePerK());
+
+    if (our_min_feerate == 0) return;
+
+    std::vector<std::pair<CNode*, CAmount>> peer_fee_filters;
+
+    size_t above_our_80_percent_fee_filter = 0;
+
+    m_connman.ForEachNode([&](CNode* pnode) {
+        if (!pnode->IsFullOutboundConn() || pnode->fDisconnect) return;
+        auto peer = GetPeerRef(pnode->GetId());
+
+        if (auto tx_relay = peer->GetTxRelay(); tx_relay != nullptr) {
+            if (tx_relay->m_fee_filter_received) {
+                peer_fee_filters.push_back(std::make_pair(pnode, tx_relay->m_fee_filter_received.load()));
+            } else {
+                ++above_our_80_percent_fee_filter;
+            }
+        }
+    });
+
+    LogPrint(BCLog::NET, "Filters? %i\n", peer_fee_filters.size() );
+
+    if (peer_fee_filters.size() == 0) return;
+
+    std::sort(peer_fee_filters.begin(), peer_fee_filters.end());
+
+    for (auto [node_id, fee_filter] : peer_fee_filters) {
+        if (fee_filter == 0) ++above_our_80_percent_fee_filter;
+        if (fee_filter >= our_min_feerate * 0.8) ++above_our_80_percent_fee_filter;;
+    }
+
+    if (above_our_80_percent_fee_filter < 4) {
+        // Drop one peer at a time to preserve a sufficient total
+        // number of connections, and to avoid network DoS.
+        // TODO add randomness chosing among low-feerate candidates.
+        peer_fee_filters[0].first->fDisconnect = true;
+    }
```

Comment:
It looks like this is backward? It seems you want to be disconnecting from peers with fee filters larger than our, but you are doing the opposite: disconnecting from the peer with the smallest fee filter (and therefore most likely the largest mempool).

Anyways, i don't think we should base this on the value of the fee filter but rather on the guesstimated size of the peer's mempool.